pkg/usage/usage.go

```diff
@@ -229,3 +230,30 @@ func (u *usage) TriggerSingleReporter(ctx context.Context) {
 		logger.Fatal(err.Error())
 	}
 }
+
+type ModelUsageHandlerParams struct {
+	UserUID, OwnerUID, ModelUID           uuid.UUID
```

Comment:
I think the requester UID should be an explicit param rather than delegating its extraction to the implementation of the usage handler. This should be the responsibility of the entrypoint or domain packages.

Reply:
Related to this, who is going to set this header?
- If this trigger comes from an `Instill` component (i.e. from a pipeline trigger), have you checked if this information is arriving? I anticipate this won't be added automatically and that we'll need to modify the component.
- Are there other ways a user could trigger a model? The public API? Should we consider setting this info anywhere else?

Reply:
I don't quite understand what it means.
Do you mean we should modify proto definitions to include the requester UID in parameters?

Reply:
No, that isn't necessary. But this is passed to the handler context as an HTTP header / gRPC metadata, so the contract doesn't need to change. 

My point is that you're extracting the user UID and the requester UID from the context, which comes from these headers, but I don't know if a trigger that comes from a pipeline is setting these. You might find that you can't subtract credit (or you're computing the wrong credit owner) because you don't have the credit owner information.

Reply:
> If this trigger comes from an Instill component (i.e. from a pipeline trigger), have you checked if this information is arriving? I anticipate this won't be added automatically and that we'll need to modify the component.

Wait, the `instill` component can't run against self-deployed models, right? Sorry, I'm not familiar yet with this component. In that case, if I understand correctly, the only entrypoint for these models is the API, right? How will it work for the models displayed in the `instill` component? Do we plan to let users specify any model they've deployed in this component? cc @heiruwu

Reply:
> Users can use any models from Instill Model in instill component, thats what the component is for

But the UI is a dropdown, which got me confused. I see in the definition that `model-name` is a string. 🤔 How does the console build the dropdown values (i.e. possible models)? cc @donch1989 

In any case, if this is so, we need to update the component to send these headers. It'd actually be a good way to pass the pipeline trigger ID and other info so in the metrics the model consumption is tied to the component consumption. But we can think about that in the future.

Reply:
on `instill.tech` you can see that I can choose the models from `Instill Model` which are accessible by me, I'm not sure what magic @donch1989 did here
![CleanShot 2024-07-04 at 21 31 25@2x](https://github.com/instill-ai/model-backend/assets/5631010/62433e88-0c59-4878-b41b-54f39c63cc7d)

> In any case, if this is so, we need to update the component to send these headers. It'd actually be a good way to pass the pipeline trigger ID and other info so in the metrics the model consumption is tied to the component consumption. But we can think about that in the future.

agree, traceability will definitely benefits us in the long run, for observability too

Reply:
cc @joremysh on the refactoring needed according to this thread

Reply:
[this](https://github.com/instill-ai/model-backend/pull/619#discussion_r1661879026), by doing something like this or other ways you see fit
> I agree, we should put the same thing as pipeline-backend in the ModelUsageHandlerParams here
// PipelineUserUID is the authenticated user executing a pipeline.
PipelineUserUID uuid.UUID `json:"__PIPELINE_USER_UID"`
// PipelineRequesterUID is the entity requesting the pipeline execution.
PipelineRequesterUID uuid.UUID `json:"__PIPELINE_REQUESTER_UID"`

Reply:
> I'm not sure what magic @donch1989 did here

🔍 From an unrelated conversation I learned that 

> The Instill Model component definition fetches the Instill Model list dynamically

That's why it was confusing

Reply:
> [this](https://github.com/instill-ai/model-backend/pull/619#discussion_r1661879026), by doing something like below or other ways you see fit
> 
> > I agree, we should put the same thing as pipeline-backend in the ModelUsageHandlerParams here
> > // PipelineUserUID is the authenticated user executing a pipeline.
> > PipelineUserUID uuid.UUID `json:"__PIPELINE_USER_UID"`
> > // PipelineRequesterUID is the entity requesting the pipeline execution.
> > PipelineRequesterUID uuid.UUID `json:"__PIPELINE_REQUESTER_UID"`

updated